File: python/238_Product_of_Array_Except_Self.py
class Solution(object):
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        ans = [1] * n
        left = 1
        for i in range(n - 1):
            left *= nums[i]
            ans[i + 1] *= left
        right = 1
        for i in range(n - 1, 0, -1):
            right *= nums[i]
            ans[i - 1] *= right
        return ans


def test():
    s = Solution()
    temp = [1, 2, 3, 4]
    print s.productExceptSelf(temp)


if __name__ == '__main__':
    test()

File: C++/238_Product_of_Array_Except_Self.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
	vector<int> productExceptSelf(vector<int>& nums) {
		int n = nums.size();
		vector<int> ans(n, 1);
		for (int i = 1; i < n; i++) {
			ans[i] = ans[i - 1] * nums[i - 1];
		}
		int right = 1;
		for (int i = n - 1; i >= 0; i--) {
			ans[i] *= right;
			right *= nums[i];
		}
		return ans;
	}
};

int main(){
	Solution s;
	int temp_array[4] = { 1, 2, 3, 4 };
	vector<int> temp(temp_array, temp_array + 4);
	vector<int> data = s.productExceptSelf(temp);
	cout << "[";
	for (int i = 0; i < data.size(); i++){
		cout << data.at(i);
		if (i == data.size() - 1)
			break;
		cout << ",";
	}
	cout << "]";

	// cin.get();
	return 0;
}